# NumPy Advanced Exercises (34–100)

_Problem-only notebook generated from the attached PDF. For each exercise, write your solution in the empty code cell that follows the problem._

**Instructions:**
- Use vectorized NumPy where possible (avoid Python loops unless asked).
- Do not modify input arrays unless the task explicitly says so.
- Add additional cells if you want experiments or tests.


In [2]:
import numpy as np

### Exercise 34

**Problem.** Create a NumPy array of all calendar dates in **July 2016** with dtype `datetime64[D]`. The range should include 2016‑07‑01 and exclude 2016‑08‑01.

In [3]:
c = np.arange('2016-07-01','2016-08-01',dtype="datetime64[D]")
c

array(['2016-07-01', '2016-07-02', '2016-07-03', '2016-07-04',
       '2016-07-05', '2016-07-06', '2016-07-07', '2016-07-08',
       '2016-07-09', '2016-07-10', '2016-07-11', '2016-07-12',
       '2016-07-13', '2016-07-14', '2016-07-15', '2016-07-16',
       '2016-07-17', '2016-07-18', '2016-07-19', '2016-07-20',
       '2016-07-21', '2016-07-22', '2016-07-23', '2016-07-24',
       '2016-07-25', '2016-07-26', '2016-07-27', '2016-07-28',
       '2016-07-29', '2016-07-30', '2016-07-31'], dtype='datetime64[D]')

### Exercise 35

**Problem.** Demonstrate **in‑place ufunc operations** using the `out=` parameter. Use at least two arrays and show one or more ufuncs that write results back into an existing array (no new temporaries).

### Exercise 36

**Problem.** Given a 1‑D array of random floats in `[0, 10)`, compute the **integer part** of each value using several different methods (e.g., arithmetic with `%`, `floor`, `ceil - 1`, `astype(int)`, `trunc`). Note differences that could arise for negative inputs (do not implement negatives here).

In [16]:
rng = np.random.default_rng(200)
r = rng.random(10)*10
r, np.floor(r), np.ceil(r) - 1, r // 1 , r.astype(int).astype(float), np.trunc(r)

(array([6.46834144, 6.63919971, 0.29901652, 1.77831342, 5.73496649,
        5.51164018, 5.26920442, 3.74242814, 0.60794097, 3.6999257 ]),
 array([6., 6., 0., 1., 5., 5., 5., 3., 0., 3.]),
 array([6., 6., 0., 1., 5., 5., 5., 3., 0., 3.]),
 array([6., 6., 0., 1., 5., 5., 5., 3., 0., 3.]),
 array([6., 6., 0., 1., 5., 5., 5., 3., 0., 3.]),
 array([6., 6., 0., 1., 5., 5., 5., 3., 0., 3.]))

### Exercise 37

**Problem.** Start with a `(5, 5)` array of zeros. **Broadcast-add** the vector `np.arange(5)` across the rows so that row `i` equals `np.arange(5)`.

In [43]:
z = np.zeros((5,5))
z + np.arange(5)

array([[0., 1., 2., 3., 4.],
       [0., 1., 2., 3., 4.],
       [0., 1., 2., 3., 4.],
       [0., 1., 2., 3., 4.],
       [0., 1., 2., 3., 4.]])

### Exercise 38

**Problem.** Build a 1‑D NumPy array **from a Python iterator/generator** that yields numbers `0..9` as floats using `np.fromiter`.

In [47]:
def gen():
    for i in range(10):
        yield i

np.fromiter(gen(), dtype="int")

array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])

### Exercise 39

**Problem.** Generate evenly spaced **interior points in the open interval (0, 1)** — i.e., exclude both endpoints — using NumPy. Return a 1‑D array.

In [5]:
np.linspace(0, 1, 10)[1:-1]

array([0.11111111, 0.22222222, 0.33333333, 0.44444444, 0.55555556,
       0.66666667, 0.77777778, 0.88888889])

### Exercise 40

**Problem.** Create a length‑10 array of random floats in `[0,1)` and **sort it in place** (mutate the original array rather than allocating a new one).

In [11]:
rng = np.random.default_rng()
x = rng.random(10)
x.sort()
x

array([0.07765824, 0.09111929, 0.16753423, 0.40928047, 0.47720368,
       0.48131256, 0.49345227, 0.65417374, 0.65663707, 0.80614614])

### Exercise 41

**Problem.** Using a vector `Z = np.arange(10)`, compute the sum via a **ufunc reduction** (e.g., `np.add.reduce(Z)`) rather than `Z.sum()`.

In [14]:
Z = np.arange(10)
np.add.reduce(Z)

np.int64(45)

### Exercise 42

**Problem.** Construct two arrays `A` and `B` and evaluate **`np.allclose(A, B)` vs `np.array_equal(A, B)`**. Return the two booleans. (Allclose tolerates tiny float differences; `array_equal` requires exact match and same shape.)

In [19]:
A=np.array([1,2,3])
B=A+1e-5
np.allclose(A,B), np.array_equal(A,B)

(True, False)

### Exercise 43

**Problem.** Create an array and make it **read‑only** (set the writeable flag to `False`). Verify that assignment would raise an error (you do not need to actually trigger an exception).

In [26]:
a = np.array([1,2,3])
a.flags.writeable = False

### Exercise 44

**Problem.** Given a `(10, 2)` array `Z` whose rows are 2‑D points `(x, y)`, compute **polar coordinates** `(R, T)` where `R = sqrt(x^2 + y^2)` and `T = arctan2(y, x)`. Return the two 1‑D arrays.

In [32]:
rng = np.random.default_rng()
Z = rng.random((10,2))
X, Y = Z[:,0], Z[:,1]
R, T = np.sqrt(X**2+Y**2), np.arctan2(Y,X)
R, T

(array([1.06850946, 0.70229693, 0.23134317, 0.21283644, 1.03352988,
        1.0130506 , 0.73704424, 1.1232301 , 0.76483624, 1.23318741]),
 array([1.20044851, 0.01729753, 0.95164315, 0.97628166, 0.32504482,
        0.68896328, 0.41805976, 1.03970455, 0.70465183, 0.81152669]))

### Exercise 45

**Problem.** Create a length‑10 random array and **replace its maximum element with 0**, in place, using vectorized indexing (no Python loops).

In [35]:
rng = np.random.default_rng()
a = rng.random(10)
a[a.argmax()] = 0
a

array([0.37188325, 0.42164849, 0.50709955, 0.53860889, 0.51986375,
       0.10318916, 0.        , 0.16413753, 0.25983461, 0.13706116])

### Exercise 46

**Problem.** Create a **structured grid** `Z` of shape `(5, 5)` with two float fields `'x'` and `'y'`. Fill each cell with its coordinate on a uniform grid over `[0,1]×[0,1]` using either `meshgrid` or broadcasting.

In [50]:
nx, ny = (5, 5)
x = np.linspace(0, 1, nx)
y = np.linspace(0, 1, ny)
xv, yv = np.meshgrid(x, y)
xv, yv

(array([[0.  , 0.25, 0.5 , 0.75, 1.  ],
        [0.  , 0.25, 0.5 , 0.75, 1.  ],
        [0.  , 0.25, 0.5 , 0.75, 1.  ],
        [0.  , 0.25, 0.5 , 0.75, 1.  ],
        [0.  , 0.25, 0.5 , 0.75, 1.  ]]),
 array([[0.  , 0.  , 0.  , 0.  , 0.  ],
        [0.25, 0.25, 0.25, 0.25, 0.25],
        [0.5 , 0.5 , 0.5 , 0.5 , 0.5 ],
        [0.75, 0.75, 0.75, 0.75, 0.75],
        [1.  , 1.  , 1.  , 1.  , 1.  ]]))

### Exercise 47

**Problem.** Let `X = np.arange(8)` and `Y = X + 0.5`. Build the **Cauchy matrix** `C` with entries `C[i, j] = 1 / (X[i] - Y[j])` using broadcasting/outer operations, then compute `det(C)`.

In [56]:
X = np.arange(8)
Y = X + 0.5
C = 1 / (X[:,None] - Y[None,:])
C[1,2] * (X[1] - Y[2]), np.linalg.det(C)

(np.float64(1.0), np.float64(3638.163637117973))

### Exercise 50

**Problem.** Given a 1‑D array `Z` of numbers (e.g., `0..99`) and a scalar `v`, return the single element of `Z` whose value is **closest to `v`**. Formally, find index `i` minimizing `|Z[i] − v|` and return `Z[i]`. Break ties by the first such element. Use **vectorized NumPy operations** (no Python loops).

In [60]:
v = 12.5
Z = np.arange(100)
np.argmin(np.abs(Z-v))

np.int64(12)

### Exercise 51

**Problem.** Create a **nested structured dtype** array of length 10 with fields `position.x`, `position.y` and `color.r`, `color.g`, `color.b` (all floats). Initialize it (values arbitrary).

In [67]:
rng = np.random.default_rng()
pos = rng.random((10,2), dtype=float)
col = rng.random((10,3), dtype=float)
np.hstack([pos, col])

array([[0.04620453, 0.54972225, 0.57552084, 0.66551153, 0.30890593],
       [0.00495369, 0.02629785, 0.99372429, 0.93325094, 0.95755553],
       [0.9785822 , 0.78784475, 0.65870582, 0.56684479, 0.79808885],
       [0.81583993, 0.30986667, 0.18738386, 0.40023684, 0.00612563],
       [0.54855062, 0.21084411, 0.32683854, 0.76344004, 0.13733922],
       [0.81105601, 0.76441386, 0.97133971, 0.14161304, 0.55153148],
       [0.71970494, 0.373309  , 0.98040169, 0.86239471, 0.91694912],
       [0.5889087 , 0.12076965, 0.83110554, 0.90884247, 0.60104391],
       [0.18480919, 0.76790804, 0.25830689, 0.88809165, 0.18667762],
       [0.7863288 , 0.15555549, 0.9786187 , 0.03032644, 0.64499372]])

### Exercise 52

**Problem.** Given `Z` of shape `(n, 2)`, compute the full **pairwise Euclidean distance matrix** `D` of shape `(n, n)` *without Python loops*, using broadcasting and reductions.

In [78]:
N = 5
rng = np.random.default_rng()
pos = rng.random((N,2), dtype=float)
D = pos[:,None,:]-pos[None,:,:]
# np.sqrt(D[:,:,0]**2+D[:,:,1]**2)
np.add.reduce(D**2,axis=2)

array([[0.        , 0.40072719, 0.574288  , 0.46459306, 0.16861741],
       [0.40072719, 0.        , 0.84199276, 0.43195526, 0.0497778 ],
       [0.574288  , 0.84199276, 0.        , 0.08221942, 0.67816288],
       [0.46459306, 0.43195526, 0.08221942, 0.        , 0.36914909],
       [0.16861741, 0.0497778 , 0.67816288, 0.36914909, 0.        ]])

### Exercise 53

**Problem.** Create a float32 array, then **reinterpret the same bytes as `int32`** using `.view()` (no copy). Show that writing the float values into the viewed array changes its integer bit patterns.

In [86]:
a = np.arange(5, dtype=np.float32)
a.view(dtype=np.int32), a

(array([         0, 1065353216, 1073741824, 1077936128, 1082130432],
       dtype=int32),
 array([0., 1., 2., 3., 4.], dtype=float32))

### Exercise 55

**Problem.** For a 2‑D array `Z` of shape `(m, n)`: (1) produce the ordered list of pairs `((i, j), Z[i, j])` for all elements; (2) produce the ordered list of all **index tuples** `(i, j)` over `{0..m-1}×{0..n-1}`. Return both lists.

In [182]:
Z = np.random.normal(0,1,6).reshape(2,3)
index = np.indices(Z.shape)
ind = np.column_stack([
    index[0].ravel(),
    index[1].ravel()
])
res = np.column_stack([ind,Z.ravel()])
res, ind

(array([[ 0.        ,  0.        ,  0.24034486],
        [ 0.        ,  1.        , -0.90233817],
        [ 0.        ,  2.        ,  1.07284402],
        [ 1.        ,  0.        ,  2.72190931],
        [ 1.        ,  1.        , -0.18510672],
        [ 1.        ,  2.        ,  0.3243922 ]]),
 array([[0, 0],
        [0, 1],
        [0, 2],
        [1, 0],
        [1, 1],
        [1, 2]]))

### Exercise 57

**Problem.** Given integers `n ≥ 1` and `0 ≤ p ≤ n^2`, create an `n×n` zero array `Z`. Using **flat indexing only** (indices `0..n^2−1`) sample `p` **distinct** indices without replacement and set those positions to 1. Return `Z`. Do **not** use 2‑D index tuples.

In [196]:
n = 3
p = 2

Z = np.zeros((n,n))
Zr = Z.ravel()
rng = np.random.default_rng()
Zr[np.random.choice(n**2, p)] = 1
Zr.reshape((n,n))

array([[0., 0., 1.],
       [0., 1., 0.],
       [0., 0., 0.]])

### Exercise 58

**Problem.** Given a matrix `X` of shape `(m, n)`, **mean‑center each row** by subtracting the per‑row mean using broadcasting (`keepdims=True`). Return the centered matrix.

In [213]:
m, n = 4, 3
X = np.random.normal(0,1,m*n).reshape(m,n)
X - np.mean(X, axis=1, keepdims=True)

array([[-2.22156287,  0.14618767,  2.0753752 ],
       [-0.60652429,  0.27331287,  0.33321142],
       [ 0.15747943, -0.27613865,  0.11865922],
       [ 0.62074592,  0.33357655, -0.95432246]])

### Exercise 59

**Problem.** Given `Z ∈ R^{m×n}` and a column index `k`, **stably sort the rows** by column `k` in ascending order using NumPy (no Python loops). Return the reordered array. Preserve relative order of ties (stable sort).

In [220]:
m, n = 4, 3
X = np.random.normal(0,1,m*n).reshape(m,n)
k = np.arange(m)
np.random.shuffle(k)
X, np.argsort(k), X[np.argsort(k)]

(array([[ 0.96463118,  0.11851392, -0.873184  ],
        [ 0.21439745, -0.82306085, -0.94567357],
        [ 0.75207097,  0.82101441, -0.59691119],
        [ 0.37103991, -1.73896576, -2.50343748]]),
 array([1, 3, 2, 0]),
 array([[ 0.21439745, -0.82306085, -0.94567357],
        [ 0.37103991, -1.73896576, -2.50343748],
        [ 0.75207097,  0.82101441, -0.59691119],
        [ 0.96463118,  0.11851392, -0.873184  ]]))

### Exercise 60

**Problem.** Given a numeric `m×n` array `Z`, decide whether there exists **any all‑zero column**. Return a single Python `bool` that is `True` iff some column is entirely zeros. Use vectorized NumPy ops only.

In [240]:
Z = np.array([
    [0,1,0,1],
    [0,0,-1,1],
    [0,1,0,1],
])
np.all(Z == 0, axis=0)

array([ True, False, False, False])

### Exercise 61

**Problem.** Given a 1‑D array `Z` of random floats and a scalar `z`, find and return the **element nearest to `z`** using the flat view (`Z.flat[...]`). Avoid Python loops.

In [247]:
Z = np.random.uniform(0, 10, 10).reshape(2, 5)
z = 7.5
ad = np.abs(Z - z)
index = np.argmin(ad)
Z.flat[index]

np.float64(8.021969807540398)

### Exercise 62

**Problem.** Let `A ∈ R^{m×1}` and `B ∈ R^{1×n}` be broadcastable arrays. Using `np.nditer` with an explicit **allocated output operand**, compute `C ∈ R^{m×n}` with entries `C[i, j] = A[i, 0] + B[0, j]`. Avoid hand‑allocating temporaries; let `nditer` allocate the output.

### Exercise 65

**Problem.** Given equal‑length arrays: integer **bin indices** `I` and **weights** `X`, compute `F` of length `1 + max(I)` where `F[k] = sum_{t} X[t] * 1[I[t] = k]`. Use vectorized NumPy (no loops). Bins with no hits should be 0.

### Exercise 66

**Problem.** Create a random `h×w×3` array of 8‑bit RGB values and compute the **number of unique colors** present. Hint: pack the three bytes per pixel into a single 24‑bit integer before `np.unique`.

### Exercise 67

**Problem.** Let `A` have shape `(..., p, q)`. Show that **summing over the last two axes** equals reshaping them into one axis and summing once: `A.sum(axis=(-2, -1)) == A.reshape(A.shape[:-2] + (-1,)).sum(axis=-1)`.

### Exercise 68

**Problem.** Given data values `D` and integer group labels `S` (same length), compute **group means** using `np.bincount` (sums and counts) and compare to `pandas` `groupby().mean()` (if available).

### Exercise 69

**Problem.** For square matrices `A, B` of the same shape, show that `diag(A @ B)` equals `sum(A * B.T, axis=1)` and also `np.einsum('ij,ji->i', A, B)`. Compute the **diagonal of a product** three ways.

### Exercise 70

**Problem.** Given a 1‑D array `Z` and integer `nz ≥ 1`, create a new array that **inserts `nz` zeros between consecutive values** of `Z`.

### Exercise 71

**Problem.** Given `A` of shape `(H, W, 3)` and `B` of shape `(H, W)`, **broadcast** `B` across the channel dimension so that each channel of `A` is scaled by `B`. Return the resulting `(H, W, 3)` array.

### Exercise 72

**Problem.** Given a 2‑D array `A`, **swap two rows** using fancy indexing (no loops, no temporary explicit copy).

### Exercise 73

**Problem.** Given a list of triangular faces as an integer array of shape `(t, 3)`, list all **undirected edges** once (deduplicate), e.g., by sorting endpoints and using `np.unique` on a structured view.

### Exercise 74

**Problem.** Given the **counts per value** as a 1‑D array `C` (e.g., from `np.bincount`), **expand** back to the full sequence by repeating each index `i` exactly `C[i]` times.

### Exercise 75

**Problem.** Implement a length‑`n` **moving average** using the cumulative sum trick so the sliding window average is computed in `O(N)` overall.

### Exercise 76

**Problem.** Using `numpy.lib.stride_tricks.as_strided`, form **rolling 1‑D windows** of length `w` from a 1‑D array as a 2‑D view of shape `(N−w+1, w)` without copies.

### Exercise 77

**Problem.** Demonstrate **in‑place logical and arithmetic flips**: flip a 0/1 integer array with `np.logical_not(..., out=...)` and flip the sign of a float array with `np.negative(..., out=...)`.

### Exercise 78

**Problem.** Given `k` line segments with endpoints `P0[i], P1[i]` in 2‑D and a query point `p`, compute the vector of **perpendicular distances** from `p` to each segment, using vectorized NumPy operations.

### Exercise 79

**Problem.** Extend the previous task: given many query points `P`, compute distances **from each point to each segment** efficiently (vectorize across points; avoid Python loops over segments).

### Exercise 80

**Problem.** Given an image array `Z` and a center `position = (row, col)`, extract a **centered crop (ROI)** of a given `shape` with **boundary checks**. Out‑of‑bounds regions should be filled with a constant `fill` value.

### Exercise 81

**Problem.** From a 1‑D array `Z`, construct a 2‑D view whose rows are all **sliding windows** of fixed width `w` (`(len(Z)−w+1, w)`), using `as_strided` (no copies).

### Exercise 82

**Problem.** Estimate the **numerical rank** of a square matrix via SVD: compute singular values and count those above a small tolerance (e.g., `1e‑10`).

### Exercise 83

**Problem.** Given an integer array, compute its **mode** using `np.bincount(...).argmax()`.

### Exercise 84

**Problem.** Given a `(H, W)` image of integers, produce a 4‑D strided view of **all 3×3 patches** with shape `((H−2), (W−2), 3, 3)` using `as_strided` (no copies).

### Exercise 85

**Problem.** Given a square matrix `Z`, **enforce symmetry** either by mirroring `Z` with `Z + Z.T − diag(diag(Z))` or by subclassing to mirror writes. Implement one approach.

### Exercise 86

**Problem.** Let `M` have shape `(p, n, n)` and `V` have shape `(p, n, 1)`. Use `np.tensordot` to perform a **tensor contraction** over the correct axes so that the result has shape `(n, 1)`.

### Exercise 87

**Problem.** Given a 2‑D array `Z` and block size `k`, compute **block sums** (coarse pooling) by reducing along both axes with `np.add.reduceat` at strides of `k`.

### Exercise 88

**Problem.** Implement one **update step of Conway’s Game of Life** on a binary grid `Z` by summing the 8 neighbors via shifted slices and applying birth/survival rules to the inner grid.

### Exercise 89

**Problem.** Given a large 1‑D array `Z` and integer `n`, extract the **top‑k elements** two ways: using `np.argsort` (exactly sorted) and `np.argpartition` (faster selection, unordered). Return both sets.

### Exercise 90

**Problem.** Write a function that computes the **Cartesian product** of several 1‑D arrays, returning an array of shape `(N, d)` containing all combinations (where `N` is the product of lengths).

### Exercise 91

**Problem.** From a 2‑D array of columns with mixed types (string/float/int), create a **record array** with named fields using `np.core.records.fromarrays`.

### Exercise 92

**Problem.** Given a large 1‑D float array `x`, compute `x**3` using **vectorized power** (`np.power(x, 3)`).

### Exercise 93

**Problem.** Given `A` of shape `(m, d)` with small integers and `B` of shape `(r, 2)`, test **row containment**: return the indices of rows `i` in `A` that contain all required values in `B` (broadcast comparisons and reduce with `any`/`all`).

### Exercise 94

**Problem.** Given a 2‑D integer array `Z`, **drop rows that are constant across all columns** (e.g., `[c, c, c, ...]`) using vectorized comparisons and boolean indexing.

### Exercise 95

**Problem.** Given a 1‑D array `I` of bytes (0..255), compute its **bit decomposition** as a 2‑D array of 0/1s with 8 columns (most‑significant bit first).

### Exercise 96

**Problem.** Given a 2‑D integer array `Z`, return the **unique rows** efficiently by viewing each row as a fixed‑width bytes object (`np.void`) and applying `np.unique` to obtain unique indices.

### Exercise 97

**Problem.** A short **einsum mini‑tour**: show examples for (a) sum of a vector, (b) elementwise product, (c) dot product, and (d) outer product, using `np.einsum` equations.

### Exercise 98

**Problem.** Given a parametric 2‑D curve `(x(t), y(t))` sampled by parameter `t`, **re‑sample uniformly by arc length**: compute cumulative chord lengths, then linearly interpolate `x` and `y` at equal arc‑length steps.

### Exercise 99

**Problem.** Given a small 2‑D float array `X`, **filter rows whose entries are all integers** and whose row‑sum equals a given integer `n`. Return the filtered subarray.

### Exercise 100

**Problem.** Given a 1‑D sample `X`, perform a **bootstrap** for the mean: draw `N` bootstrap resamples (with replacement), compute the resample means, and report a two‑sided percentile **confidence interval** (e.g., 2.5% and 97.5%).